wgpu-hal/src/lib.rs

```diff
@@ -190,50 +191,196 @@ impl InstanceError {
     }
 }
 
-pub trait Api: Clone + fmt::Debug + Sized {
-    type Instance: Instance<A = Self>;
-    type Surface: Surface<A = Self>;
-    type Adapter: Adapter<A = Self>;
-    type Device: Device<A = Self>;
+macro_rules! resource {
+    ($name:ident $(: $($bounds:tt)*)?) => {
+        pub trait $name: Resource $(+ $($bounds)*)? {}
+        impl<T: Resource $(+ $($bounds)*)?> $name for T {}
+    };
+}
 
-    type Queue: Queue<A = Self>;
-    type CommandEncoder: CommandEncoder<A = Self>;
-    type CommandBuffer: WasmNotSendSync + fmt::Debug;
+pub trait Resource: WasmNotSendSync + 'static {
+    fn as_any(&self) -> &dyn Any;
+    fn as_any_mut(&mut self) -> &mut dyn Any;
+}
+impl<T: WasmNotSendSync + 'static> Resource for T {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+}
 
-    type Buffer: fmt::Debug + WasmNotSendSync + 'static;
-    type Texture: fmt::Debug + WasmNotSendSync + 'static;
-    type SurfaceTexture: fmt::Debug + WasmNotSendSync + Borrow<Self::Texture>;
-    type TextureView: fmt::Debug + WasmNotSendSync;
-    type Sampler: fmt::Debug + WasmNotSendSync;
-    type QuerySet: fmt::Debug + WasmNotSendSync;
-    type Fence: fmt::Debug + WasmNotSendSync;
+resource!(CommandBuffer: fmt::Debug);
 
-    type BindGroupLayout: fmt::Debug + WasmNotSendSync;
-    type BindGroup: fmt::Debug + WasmNotSendSync;
-    type PipelineLayout: fmt::Debug + WasmNotSendSync;
-    type ShaderModule: fmt::Debug + WasmNotSendSync;
-    type RenderPipeline: fmt::Debug + WasmNotSendSync;
-    type ComputePipeline: fmt::Debug + WasmNotSendSync;
+resource!(Buffer: fmt::Debug + 'static);
+resource!(Texture: fmt::Debug + 'static);
+resource!(TextureView: fmt::Debug);
+resource!(Sampler: fmt::Debug);
+resource!(QuerySet: fmt::Debug);
+resource!(Fence: fmt::Debug);
 
-    type AccelerationStructure: fmt::Debug + WasmNotSendSync + 'static;
-}
+resource!(BindGroupLayout: fmt::Debug);
+resource!(BindGroup: fmt::Debug);
+resource!(PipelineLayout: fmt::Debug);
+resource!(ShaderModule: fmt::Debug);
+resource!(RenderPipeline: fmt::Debug);
+resource!(ComputePipeline: fmt::Debug);
+
+resource!(AccelerationStructure: fmt::Debug + 'static);
 
-pub trait Instance: Sized + WasmNotSendSync {
-    type A: Api;
+pub trait SurfaceTexture<T: Texture + ?Sized>: Resource + fmt::Debug + Borrow<T> {}
+impl<T: Resource + fmt::Debug + Borrow<Tex>, Tex: Texture> SurfaceTexture<Tex> for T {}
 
-    unsafe fn init(desc: &InstanceDescriptor) -> Result<Self, InstanceError>;
+pub trait Api: Clone + fmt::Debug + Sized {
```

Comment:
there should be some docs on how `Api` is different from `ConcreteApi` and what's the purpose of the two

Reply:
Done.